crates/derive/src/types/batch/span_batch/bits.rs

```diff
@@ -0,0 +1,187 @@
+//! Module for working with span batch bits.
+
+use crate::types::{SpanBatchError, MAX_SPAN_BATCH_SIZE};
+use alloc::{vec, vec::Vec};
+use alloy_rlp::Buf;
+use anyhow::Result;
+
+/// Type for span batch bits.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct SpanBatchBits(pub Vec<u8>);
+
+impl AsRef<Vec<u8>> for SpanBatchBits {
+    fn as_ref(&self) -> &Vec<u8> {
+        &self.0
+    }
+}
+
+impl AsRef<[u8]> for SpanBatchBits {
+    fn as_ref(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+impl From<SpanBatchBits> for Vec<u8> {
+    fn from(bits: SpanBatchBits) -> Vec<u8> {
+        bits.0
+    }
+}
+
+impl SpanBatchBits {
+    /// Decodes a standard span-batch bitlist from a reader.
+    /// The bitlist is encoded as big-endian integer, left-padded with zeroes to a multiple of 8 bits.
+    /// The encoded bitlist cannot be longer than [MAX_SPAN_BATCH_SIZE].
+    pub fn decode(b: &mut &[u8], bit_length: usize) -> Result<Self, SpanBatchError> {
+        let buffer_len = bit_length / 8 + if bit_length % 8 != 0 { 1 } else { 0 };
+        if buffer_len > MAX_SPAN_BATCH_SIZE {
+            return Err(SpanBatchError::TooBigSpanBatchSize);
+        }
+
+        // TODO(refcell): This can definitely be optimized.
+        let bits = if b.len() < buffer_len {
+            let mut bits = vec![0; buffer_len];
+            bits[..b.len()].copy_from_slice(b);
+            b.advance(b.len());
+            bits
+        } else {
+            let v = b[..buffer_len].to_vec();
+            b.advance(buffer_len);
+            v
+        };
+        let sb_bits = SpanBatchBits(bits.to_vec());
+
+        // TODO(clabby): Why doesn't this check work?
+        // if sb_bits.bit_len() > bit_length {
+        //     return Err(SpanBatchError::BitfieldTooLong);
+        // }
```

Comment:
I think this is an incorrect check - the bitlist is left-padded with zeros using the `buffer_len` variable.

The `bit_length` variable is not padded to a multiple of 8 bits so the `sb_bits.bit_len()` can be larger that than the `bit_length` variable and be valid.

I believe the correct check should be:
```suggestion
        if sb_bits.bit_len() > (buffer_len * 8) {
        	return Err(SpanBatchError::BitfieldTooLong);
        }
```